### PR TITLE
Make Imp tests setup more realistic

### PR DESCRIPTION
**Pull request by lehins**
# Description

In this PR we are switching to the same way that era is setup in real life. Namely by taking the state of the previous era together with the genesis data and constructing the ledger state for the new era.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 1]
```diff
@@ -1603,3 +1628,49 @@ submitYesVoteCCs_ ::
   ImpTestM era ()
 submitYesVoteCCs_ committeeMembers govId =
   mapM_ (\c -> submitYesVote_ (CommitteeVoter c) govId) committeeMembers
+
+submitUpdateCommittee ::
```
Comment (teodanciu) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1632]: I think the name can create confusion, because `UpdateCommittee` action specifies both members to remove and to add. So this list could be interpreted as the final desired committee.  How about `submitAddCommitteeMembers` ? 

```suggestion
submitAddCommitteeMembers ::
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 2]
```diff
@@ -1603,3 +1628,49 @@ submitYesVoteCCs_ ::
...
+      Map.fromList [(cc, addEpochInterval curEpochNo lifetime) | (cc, lifetime) <- ccs]
+  submitGovAction $ UpdateCommittee parent mempty newCommitteMembers threshold
+
+expectCommitteeMemberPresence ::
+  (HasCallStack, ConwayEraGov era) => Credential 'ColdCommitteeRole (EraCrypto era) -> ImpTestM era ()
+expectCommitteeMemberPresence cc = do
+  SJust committee <-
+    getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . committeeGovStateL
+  assertBool ("Expected Committee Member: " ++ show cc ++ " to be present in the committee") $
+    Map.member cc (committee ^. committeeMembersL)
```
Comment (teodanciu) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1659]: We could reuse `getCommitteeMembers` to avoid duplication: 
```suggestion
  (HasCallStack, ConwayEraImp era) => Credential 'ColdCommitteeRole (EraCrypto era) -> ImpTestM era ()
expectCommitteeMemberPresence cc = do
  members <- getCommitteeMembers
  assertBool ("Expected Committee Member: " ++ show cc ++ " to be present in the committee") $
    Set.member cc member
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 3]
```diff
@@ -1603,3 +1628,49 @@ submitYesVoteCCs_ ::
...
+
+expectCommitteeMemberAbsence ::
+  (HasCallStack, ConwayEraGov era) => Credential 'ColdCommitteeRole (EraCrypto era) -> ImpTestM era ()
+expectCommitteeMemberAbsence cc = do
+  SJust committee <-
+    getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . committeeGovStateL
+  assertBool ("Expected Committee Member: " ++ show cc ++ " to be absent from the committee") $
+    Map.notMember cc (committee ^. committeeMembersL)
```
Comment (teodanciu) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1667]: Same as before about reusing `getCommitteeMembers`: 
```suggestion
  (HasCallStack, ConwayEraImp era) => Credential 'ColdCommitteeRole (EraCrypto era) -> ImpTestM era ()
expectCommitteeMemberAbsence cc = do
  members <- getCommitteeMembers
  assertBool ("Expected Committee Member: " ++ show cc ++ " to be absent from the committee") $
    Set.notMember cc members
```

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs [hunk 4]
```diff
@@ -45,83 +41,105 @@ import Test.Cardano.Ledger.Imp.Common
...
-        let result1 = queryDRepState nes1 mempty
-        let state1 = fromJust $ Map.lookup drep result1
-        state1 ^. drepExpiryL `shouldBe` EpochNo 5
+        drepState1 <- drepStateFromQuery drep $ queryDRepState nes1 mempty
+        drepState1
+          ^. drepExpiryL
+            `shouldBe` addEpochInterval
+              curEpochNo
+              (EpochInterval (drepActivity + fromIntegral n))
         passNEpochsChecking 3 $
```
Comment (teodanciu) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs]: I can't tell for sure whether the `3` is in relation to `drepActivity`, `n` or it's not actually related.   If it's the latter, please ignore this, otherwise  we could use the variable, for more explicit meaning, since it's defined: 

```suggestion
        passNEpochsChecking (fromIntegral drepActivity) $
```

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs [hunk 5]
```diff
@@ -45,83 +41,105 @@ import Test.Cardano.Ledger.Imp.Common
...
 spec = do
   describe "DRep" $ do
     describe "Expiries are reported correctly" $ do
+      let drepStateFromQuery ::
+            (HasCallStack, Monad m) =>
+            Credential 'DRepRole (EraCrypto era) ->
+            Map.Map (Credential 'DRepRole (EraCrypto era)) (DRepState (EraCrypto era)) ->
+            m (DRepState (EraCrypto era))
+          drepStateFromQuery drep queryResult =
+            case Map.lookup drep queryResult of
```
Comment (teodanciu) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs]: I find it a little odd to call it `drepStateFromQuery` and to take a Map as a parameter. 
Why not run the query within this function? I know it's less flexible, but it seems like it's used always in the same way: 

```suggestion
            NewEpochState era ->
            m (DRepState (EraCrypto era))
          drepStateFromQuery drep queryResult =
            case Map.lookup drep (queryDRepState nes mempty) of
```

Or if we want to keep the query execution outside, we could rename the function to `lookupDrepState`

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs [hunk 6]
```diff
@@ -45,83 +41,105 @@ import Test.Cardano.Ledger.Imp.Common
...
         nes2 <- getsNES id
-        let result2 = queryDRepState nes2 mempty
-        let state2 = fromJust $ Map.lookup drep result2
-        state2 ^. drepExpiryL `shouldBe` EpochNo 7
-        expectActualDRepExpiry drep 7
+        drepState2 <- drepStateFromQuery drep $ queryDRepState nes2 mempty
+        let drepExpiry2 = addEpochInterval curEpochNo $ EpochInterval (drepActivity + fromIntegral n + 1)
+        drepState2 ^. drepExpiryL `shouldBe` drepExpiry2
+        expectActualDRepExpiry drep drepExpiry2
         passNEpochsChecking 3 $ do
```
Comment (teodanciu) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs]: Same question as before, would this make sense? : 
```suggestion
        passNEpochsChecking (fromIntegral drepActivity) $ do
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs [hunk 7]
```diff
@@ -214,7 +214,7 @@ basicShelleyTxBodyRaw =
     { stbrInputs = Set.empty
     , stbrOutputs = StrictSeq.empty
     , stbrTxFee = Coin 0
-    , stbrTTL = SlotNo 0
+    , stbrTTL = SlotNo maxBound -- transaction is eternally valid by default
```
Comment (teodanciu) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:217]: Just curious, why this change? Is there no potential for problems since we are loosening in a way the conditions?

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 8]
```diff
@@ -448,101 +612,62 @@ logStakeDistr = do
...
+              & ppEMaxL .~ EpochInterval 18
+              & ppNOptL .~ 150
+              & ppA0L .~ (3 %! 10)
+              & ppRhoL .~ (3 %! 1000)
+              & ppTauL .~ (2 %! 10)
+              & ppDL .~ (1 %! 1)
+              & ppExtraEntropyL .~ NeutralNonce
+              & ppMinUTxOValueL .~ Coin 2_000_000
+              & ppMinPoolCostL .~ Coin 340_000_000
+        , -- TODO: Add a top level definition and add pricate keys to ImpState:
```
Comment (teodanciu) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]:
```suggestion
        , -- TODO: Add a top level definition and add private keys to ImpState:
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 9]
```diff
@@ -934,14 +1086,24 @@ fixupFees tx = impAnn "fixupFees" $ do
             & bodyTxL . feeTxBodyL .~ (fee <> change)
   pure txWithFee
 
+-- | Adds an auxiliary data hash if auxiliary data present, while and the hash of it is not.
```
Comment (teodanciu) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]:
```suggestion
-- | Adds an auxiliary data hash if auxiliary data present, while the hash of it is not.
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 10]
```diff
@@ -885,11 +1020,28 @@ impNativeScriptKeyPairs tx = do
...
+    pure txOut'
+  pure $ tx & bodyTxL . outputsTxBodyL .~ fixedUpTxOuts
+
 fixupFees ::
   (ShelleyEraImp era, HasCallStack) =>
   Tx era ->
   ImpTestM era (Tx era)
-fixupFees tx = impAnn "fixupFees" $ do
+fixupFees txOriginal = impAnn "fixupFees" $ do
+  tx <- fixupTxOuts txOriginal
```
Comment (teodanciu) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Just curious, did the change in initializing from genesis made this necessary? Is it because of the removal of `addRootTxOut` ?

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 11]
```diff
@@ -934,14 +1086,24 @@ fixupFees tx = impAnn "fixupFees" $ do
...
+  | otherwise = pure tx
+
 shelleyFixupTx ::
   forall era.
   (ShelleyEraImp era, HasCallStack) =>
   Tx era ->
   ImpTestM era (Tx era)
 shelleyFixupTx =
   addNativeScriptTxWits
+    >=> fixupAuxDataHash
```
Comment (teodanciu) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:1113]: Same question as above - have the rest of the changes made this necessary?

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs (hunk 7 above)
Comment (teodanciu) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:217]: Would it be good to move it in a separate commit?

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxoSpec.hs [hunk 12]
```diff
@@ -20,10 +21,13 @@ spec ::
   ) =>
   SpecWith (ImpTestState era)
 spec = describe "UTXO" $ do
+  it "sendCoinTo" $ do
+    addr <- freshKeyAddr_
+    void $ sendCoinTo addr (Coin 999833135)
```
Comment (teodanciu) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxoSpec.hs]: What's the point of this test? Just checking that `sendCoinTo` works without error? Is this a significant value?

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 11 above)
Comment (teodanciu) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:1113]: Ah is it maybe related to the test that's blowing up the fee to increase the collateral?

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxoSpec.hs [hunk 13]
```diff
@@ -52,14 +55,23 @@ spec = describe "UTXO" $ do
...
+          -- including random garbage. Auxiliary data fits the bill quite nicely
+          metadata <-
+            Map.fromList
+              <$> forM [1 .. (12 * 1024 `div` 64)] (\ix -> (,) ix . M.B <$> uniformByteStringM 64)
+          let auxData = mkAlonzoTxAuxData @[] @era metadata []
+              tx =
                 mkBasicTx mkBasicTxBody
                   & bodyTxL . inputsTxBodyL <>~ [scriptInput]
                   & bodyTxL . collateralInputsTxBodyL <>~ [collateralInput]
+                  & auxDataTxL .~ SJust auxData
```
Comment (teodanciu) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxoSpec.hs:74]: This causes `MissingTxBodyMetadataHash` in `UtxowSpec` to fail. 
But you can fix it easily by adding these  two lines before [submitFailingTx](https://github.com/IntersectMBO/cardano-ledger/blob/lehins/various-fixes-to-imptests/eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxowSpec.hs#L92): 

```
    let resetAddrWits = updateAddrTxWits . (witsTxL . addrTxWitsL .~ [])
    withPostFixup (resetAddrWits . (bodyTxL . auxDataHashTxBodyL .~ SNothing )) $
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovCertSpec.hs [hunk 14]
```diff
@@ -60,12 +35,27 @@ import Test.Cardano.Ledger.Imp.Common
...
+    passNEpochs n
+    (drepCred, _, _) <- setupSingleDRep 1_000_000
+    cc <- KeyHashObj <$> freshKeyHash
+    EpochInterval committeeMaxTermLength <-
+      getsNES $ nesEsL . curPParamsEpochStateL . ppCommitteeMaxTermLengthL
+    secondAddCCGaid <-
+      submitUpdateCommittee Nothing [(cc, EpochInterval (committeeMaxTermLength + 2))] (1 %! 2)
+    submitYesVote_ (DRepVoter drepCred) secondAddCCGaid
+    passNEpochs 2
+    -- Due to longer than expected lifetime we have to wait an extra epoch for this new action to be enacted
```
Comment (teodanciu) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovCertSpec.hs]:
```suggestion
    -- Due to longer than allowed lifetime we have to wait an extra epoch for this new action to be enacted
```

**Review by teodanciu — APPROVED** [review]
Looks good to me, just a few comments and questions, and a suggestion to fix a failing test.
Great to have the Imp initialization given by the genesis mechanism, instead of some other ad-hoc structure!

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs (hunk 1 above)
Comment (lehins) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1632]: It is the name of the action that is important here: `UpdateCommittee`. It just looks like we don't have a single test that verifies removal of committee members, that's why this function used `mempty` for CC members to be removed. I've added an extra argument, kept the name and created a ticket: https://github.com/IntersectMBO/cardano-ledger/issues/4527

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs (hunk 2 above)
Comment (lehins) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1659]: I actually thought about it. There are two downsides to this:

1. When no committee is present it will result in a different failure (pattern match on `SNothing`), which will make it clear that we are in the state of no confidence
2. Much less important for testing, but we would do a redundant conversion to a Set.

Comment (lehins) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1659]: In other words there is a difference between an empty comittee and no committee and `getCommitteeMembers` erases that difference

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs (hunk 3 above)
Comment (lehins) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:1667]: Same response as before

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs (hunk 4 above)
Comment (lehins) [reply libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs]: Yeah, I think you are right. It seems to be checking that the DRep did not expire throughout the whole period of `drepActivity`. I've applied your suggestion manually

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs (hunk 5 above)
Comment (lehins) [reply libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs]: Yep, it is always used in the same way. Nice simplification. Thanks.

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs (hunk 6 above)
Comment (lehins) [reply libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/Imp/QuerySpec.hs]: Same response. You are absolutely right.

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxoSpec.hs (hunk 13 above)
Comment (lehins) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxoSpec.hs:74]: I was confused for a second how this test could be causing another test to fail, but looking further into the test failure I realized that you meant the `fixupAuxDataHash` that was needed for this test is causing the `MissingTxBodyMetadataHash` to fail.

Unfortunately, your suggestion will not work, since removing witnesses will cause the transaction to fail, cause signatures for the inputs being spent will be missing. We can't change just the `auxDataHashTxBodyL` after fixup either, because it will affect signatures.

A much easier fix to this is to add the actual `auxData` after the fixup :slightly_smiling_face: 

```haskell
        tx = mkBasicTx $ mkBasicTxBody & feeTxBodyL .~ Coin 165545
        addAuxData fixedTx = pure (fixedTx & auxDataTxL .~ SJust auxData)
    withPostFixup addAuxData $
```

One caveat is that the fee must be set manually, since addition of aux data affects the size of the transaction consequently making the fixud up fee insufficient. I've also added functionality that prevents mocking with the `feeTxBodyL` field if it is already present.

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxoSpec.hs (hunk 12 above)
Comment (lehins) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxoSpec.hs]: It was a temporary test that I was using it to implement MinUTxO fixup. It is no longer needed, since as you very well pointed out, there is little value in having it.

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 11 above)
Comment (lehins) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:1113]: That's right. Instead of manually supplying the hash for auxiliary data, I decided to create a general fixup function that will take care of it automatically

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 10 above)
Comment (lehins) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: This fixup is one of the biggest changes in this PR, which we need because genesis file forces minutxo value for all outputs. For this reason none of those `sendCoinTo addr (Coin 10)` will work, simply because the amount being sent is now too small.

Thinking a bit more about this function after your review I decided to change it a bit, namely amount will be fixed up only if it is originally set to `0`. This way it will be a lot less surprising when writing tests.

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs (hunk 7 above)
Comment (lehins) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:217]: The reason for this change is to make it consistent with Allegra onwards, which makes transactions with valid and permissive interval by default: https://github.com/IntersectMBO/cardano-ledger/blob/411054e40b4e08350049e4eaffdf0cc5f73e9d91/eras/allegra/impl/src/Cardano/Ledger/Allegra/TxBody.hs#L197

> Is there no potential for problems since we are loosening in a way the conditions?

Not really, since no-one constructs Shelley transactions anymore, except the core team for testing.

It was already in a separate commit 84424186558e45e9dff5419c47d9b99db0c3b590

**Comment by teodanciu** [comment]
I don't know what's going on with building for 8.10.7. The job keeps failing but I don't see logs . Seems to be failing when reaching "Archive working directory" step -  maybe again the size thing as we had before?
